bofire/strategies/doe_strategy.py

```diff
@@ -20,22 +37,139 @@ def __init__(
+        new_domain, new_categories = discrete_to_relaxable_domain_mapper(self.domain)
+        all_new_categories.extend(new_categories)
+
+        # check for NchooseK constraint and solve the problem differently depending on the strategy
+        if self.data_model.optimization_strategy != "partially-random":
+            (
+                new_domain,
+                new_categories,
+            ) = nchoosek_to_relaxable_domain_mapper(new_domain)
+            all_new_categories.extend(new_categories)
+
+        # check categorical_groups
+        validate_categorical_groups(all_new_categories, new_domain)
+
+        # here we adapt the (partially) fixed experiments to the new domain
+        # todo
+
         if self.candidates is not None:
             _fixed_experiments_count = len(self.candidates)
             _candidate_count = candidate_count + len(self.candidates)
         else:
             _fixed_experiments_count = 0
             _candidate_count = candidate_count
 
-        design = find_local_max_ipopt(
-            self.domain,
-            self.formula,
-            n_experiments=_candidate_count,
-            fixed_experiments=self.candidates,
+        num_binary_vars = len(new_domain.get_features(includes=[RelaxableBinaryInput]))
+        num_discrete_vars = len(
+            new_domain.get_features(includes=[RelaxableDiscreteInput])
         )
-        return design.iloc[_fixed_experiments_count:, :]  # type: ignore
+
+        if (
+            self.data_model.optimization_strategy == "relaxed"
+            or (num_binary_vars == 0 and num_discrete_vars == 0)
+            or (
+                self.data_model.optimization_strategy == "partially-random"
+                and num_binary_vars == 0
+                and num_discrete_vars == 0
+            )
+        ):
+            design = find_local_max_ipopt(
+                new_domain,
+                self.formula,
+                n_experiments=_candidate_count,
+                fixed_experiments=self.candidates,
+                partially_fixed_experiments=self._partially_fixed_experiments_for_next_design,
+            )
+        # todo adapt to when exhaustive search accepts discrete variables
+        elif (
+            self.data_model.optimization_strategy == "exhaustive"
+            and num_discrete_vars == 0
+        ):
+            design = find_local_max_ipopt_exhaustive(
+                domain=new_domain,
+                model_type=self.formula,
+                n_experiments=_candidate_count,
+                fixed_experiments=self.candidates,
+                verbose=self.data_model.verbose,
+                partially_fixed_experiments=self._partially_fixed_experiments_for_next_design,
+                categorical_groups=all_new_categories,
+            )
+        elif self.data_model.optimization_strategy in [
+            "branch-and-bound",
+            "default",
+            "partially-random",
```

Comment:
why do we have again all these strategies that map to the same optimizer?

Reply:
depending on which Inputs are in the domain, need to use the BaB regardless of which strategy is provided. For example, if there are CategoricalInputs and NChooseKConstraints and the strategy is set to partially-random. We need solve the NChosseKConstraints by using the already existing strategy where we randomly sample solution, but the CategoricalInputs have to be solved with the BaB.

Reply:
Ok.  Can you please add comments that explain these cases and what the strategies mean? And maybe "partially-random" is not very idiomatic. It is referring to  the current NChooseKConstraint implementation, correct? Maybe call it "random-nchoosek-realization"?